File: app/src/main/java/com/denchic45/kts/data/model/mapper/CourseMapper.java
package com.denchic45.kts.data.model.mapper;

import androidx.annotation.NonNull;

import com.denchic45.kts.data.model.domain.Course;
import com.denchic45.kts.data.model.domain.CourseInfo;
import com.denchic45.kts.data.model.domain.Group;
import com.denchic45.kts.data.model.firestore.CourseDoc;
import com.denchic45.kts.data.model.room.CourseEntity;
import com.denchic45.kts.data.model.room.CourseWithSubjectAndTeacher;
import com.denchic45.kts.data.model.room.CourseWithSubjectWithTeacherAndGroups;
import com.denchic45.kts.utils.SearchKeysGenerator;

import org.jetbrains.annotations.NotNull;
import org.mapstruct.AfterMapping;
import org.mapstruct.InheritInverseConfiguration;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.Named;

import java.util.List;
import java.util.stream.Collectors;

@Mapper(uses = {GroupMapper.class, UserMapper.class, SubjectMapper.class})
public interface CourseMapper {

    @Mapping(source = "info.teacher", target = "teacher")
    @Mapping(source = "info.subject", target = "subject")
    @Mapping(source = "info.name", target = "name")
    @Mapping(qualifiedByName = "addGroupId", source = "groups", target = "groupIds")
    CourseDoc domainToDoc(Course course);

    CourseInfo docToDomain(CourseDoc doc);

    List<CourseInfo> docToDomain(List<CourseDoc> docs);

    @Mapping(source = "groupEntities", target = "groups")

//    @Mapping(source = "courseEntity.name", target = "info.name")
//    @Mapping(source = "courseEntity.uuid", target = "info.uuid")
//    @Mapping(source = "subjectEntity", target = "info.subject")
//    @Mapping(source = "teacherEntity", target = "info.teacher")
    @Mapping(source = ".", target = "info", qualifiedByName = "entityToDomainInfo")
    Course entityToDomain(CourseWithSubjectWithTeacherAndGroups entity);

    @Mapping(source = "subject.id", target = "subjectId")
    @Mapping(source = "teacher.id", target = "teacherId")
    CourseEntity docToEntity(CourseDoc doc);

    List<CourseEntity> docToEntity(List<CourseDoc> doc);

    @InheritInverseConfiguration(name = "docToEntity")
    CourseDoc entityToDoc(CourseEntity entity);

    List<Course> entityToDomain(List<CourseWithSubjectWithTeacherAndGroups> entities);

    @Mapping(source = "courseEntity", target = ".")
    @Named("entityToDomainInfo")
    @Mapping(source = "subjectEntity", target = "subject")
    @Mapping(source = "teacherEntity", target = "teacher")
    CourseInfo entityToDomainInfo(CourseWithSubjectWithTeacherAndGroups entity);

    @Mapping(source = "courseEntity", target = ".")
    @Named("entityToDomainInfo")
    @Mapping(source = "subjectEntity", target = "subject")
    @Mapping(source = "teacherEntity", target = "teacher")
    CourseInfo entityToDomainInfo(CourseWithSubjectAndTeacher entity);

    default List<CourseInfo> entityToDomainInfo(@NonNull List<CourseWithSubjectWithTeacherAndGroups> entities) {
        return entities.stream().map(this::entityToDomainInfo).collect(Collectors.toList());
    }

    default List<CourseInfo> entityToDomainInfo2(@NonNull List<CourseWithSubjectAndTeacher> entities) {
        return entities.stream().map(this::entityToDomainInfo).collect(Collectors.toList());
    }

    @Named("addGroupId")
    default List<String> addGroupId(@NonNull List<Group> groups) {
        return groups.stream().map(Group::getId).collect(Collectors.toList());
    }

    @AfterMapping
    default void addSearchKeys(@MappingTarget @NotNull CourseDoc courseDoc) {
        SearchKeysGenerator generator = new SearchKeysGenerator();
        courseDoc.setSearchKeys(generator.generateKeys(courseDoc.getName(), predicate -> predicate.length() > 1));
    }
}

File: app/src/main/java/com/denchic45/kts/data/model/mapper/CourseMapper.kt
package com.denchic45.kts.data.model.mapper

import com.denchic45.kts.domain.model.Course
import com.denchic45.kts.domain.model.GroupHeader
import com.denchic45.kts.domain.model.CourseHeader
import com.denchic45.kts.data.remote.model.CourseDoc
import com.denchic45.kts.data.model.room.CourseEntity
import com.denchic45.kts.data.model.room.CourseWithSubjectAndTeacherEntities
import com.denchic45.kts.data.model.room.CourseWithSubjectWithTeacherAndGroupsEntities
import com.denchic45.kts.data.model.room.GroupWithCuratorAndSpecialtyEntity
import org.mapstruct.*

@Mapper(uses = [GroupMapper::class, UserMapper::class, SubjectMapper::class, SpecialtyMapper::class])
abstract class CourseMapper {

    @Mapping(qualifiedByName = ["courseGroupToGroupId"], source = "groupHeaders", target = "groupIds")
    abstract fun domainToDoc(course: Course): CourseDoc

    abstract fun docToDomain(doc: CourseDoc): Course

    abstract fun docToDomain(docs: List<CourseDoc>): List<CourseHeader>

    abstract fun docToDomain2(docs: List<CourseDoc>): List<CourseHeader>

    @Named("groupWithCuratorAndSpecialtyEntityToCourseGroup")
    @Mapping(source = "entities.groupEntity", target = ".")
    abstract fun groupWithCuratorAndSpecialtyEntityToCourseGroup(entities: GroupWithCuratorAndSpecialtyEntity): GroupHeader

    @IterableMapping(qualifiedByName = ["groupWithCuratorAndSpecialtyEntityToCourseGroup"])
    abstract fun groupWithCuratorAndSpecialtyEntityToCourseGroup(entities: List<GroupWithCuratorAndSpecialtyEntity>): List<GroupHeader>

    @Mapping(source = "groupEntities", target = "groupHeaders")
    @Mapping(source = "courseEntity.name", target = "name")
    @Mapping(source = "courseEntity.id", target = "id")
    @Mapping(source = "subjectEntity", target = "subject")
    @Mapping(source = "teacherEntity", target = "teacher")
    @Named("entityToDomain2")
    abstract fun entityToDomain2(entity: CourseWithSubjectWithTeacherAndGroupsEntities): Course

    @IterableMapping(qualifiedByName = ["entityToDomain2"])
    abstract fun entityToDomain2(entity: List<CourseWithSubjectWithTeacherAndGroupsEntities>): List<Course>

    @Mapping(source = "courseEntity.name", target = "name")
    @Mapping(source = "courseEntity.id", target = "id")
    @Mapping(source = "subjectEntity", target = "subject")
    @Mapping(source = "teacherEntity", target = "teacher")
    @Named("entityToDomainHeaders")
    abstract fun entityToDomainHeaders(entity: CourseWithSubjectAndTeacherEntities): CourseHeader

    @IterableMapping(qualifiedByName = ["entityToDomainHeaders"])
    abstract fun entityToDomainHeaders(entity: List<CourseWithSubjectAndTeacherEntities>): List<CourseHeader>

    @Mapping(source = "subject.id", target = "subjectId")
    @Mapping(source = "teacher.id", target = "teacherId")
    abstract fun docToEntity(doc: CourseDoc): CourseEntity

    abstract fun docToEntity(doc: List<CourseDoc>): List<CourseEntity>

    abstract fun entityToDomain(entities: List<CourseWithSubjectAndTeacherEntities>): List<Course>

    @Mapping(source = "teacherEntity", target = "teacher")
    @Mapping(source = "subjectEntity", target = "subject")
    @Mapping(source = "courseEntity", target = ".")
    abstract fun entityToDoc(entities: CourseWithSubjectAndTeacherEntities): CourseDoc

    @Named("courseGroupToGroupId")
    fun courseGroupToGroupId(groupHeader: GroupHeader): String {
        return groupHeader.id
    }


}